File: ui/src/config/eq.ts
/**
 * A parametric EQ sitting between the analyser and the colour mapping.
 *
 * The engine works on band magnitudes, not samples, so the EQ is applied as a
 * gain curve in dB rather than by running audio through a filter. That makes
 * the *shape* of the curve the whole specification — and the shape people
 * expect from a parametric EQ is the RBJ cookbook biquad, so that is what is
 * evaluated here rather than an invented bell.
 *
 * Doing it properly is not much more code than faking it with a Gaussian, and
 * it means the engine can implement this with the same coefficients and get the
 * same answer instead of being reverse-engineered from a drawing.
 */

export type EqType = "peak" | "lowShelf" | "highShelf" | "lowPass" | "highPass";

export interface EqBand {
  id: string;
  type: EqType;
  hz: number;
  /** dB. Ignored by the pass filters, which cut rather than gain. */
  gain: number;
  q: number;
}

/** Vertical extent of the gain scale drawn over the plot. */
export const EQ_RANGE_DB = 24;

export const EQ_Q_MIN = 0.1;
export const EQ_Q_MAX = 12;

export const EQ_TYPE_OPTIONS: Array<{ value: EqType; label: string }> = [
  { value: "peak", label: "Bell" },
  { value: "lowShelf", label: "Low shelf" },
  { value: "highShelf", label: "High shelf" },
  { value: "lowPass", label: "Low pass" },
  { value: "highPass", label: "High pass" },
];

/** Pass filters have no gain, so their node is pinned to the zero line. */
export function hasGain(type: EqType): boolean {
  return type === "peak" || type === "lowShelf" || type === "highShelf";
}

export const DEFAULT_SAMPLE_RATE = 48_000;

interface Biquad {
  b0: number;
  b1: number;
  b2: number;
  a0: number;
  a1: number;
  a2: number;
}

/** RBJ Audio EQ Cookbook, verbatim. */
function coefficients(band: EqBand, fs: number): Biquad {
  const nyquist = fs / 2;
  const f0 = Math.min(Math.max(band.hz, 1), nyquist * 0.999);
  const q = Math.min(Math.max(band.q, EQ_Q_MIN), EQ_Q_MAX);
  const w0 = (2 * Math.PI * f0) / fs;
  const cos = Math.cos(w0);
  const sin = Math.sin(w0);
  const alpha = sin / (2 * q);
  const A = Math.pow(10, band.gain / 40);
  const sqrtA = Math.sqrt(A);

  switch (band.type) {
    case "peak":
      return {
        b0: 1 + alpha * A,
        b1: -2 * cos,
        b2: 1 - alpha * A,
        a0: 1 + alpha / A,
        a1: -2 * cos,
        a2: 1 - alpha / A,
      };
    case "lowShelf":
      return {
        b0: A * (A + 1 - (A - 1) * cos + 2 * sqrtA * alpha),
        b1: 2 * A * (A - 1 - (A + 1) * cos),
        b2: A * (A + 1 - (A - 1) * cos - 2 * sqrtA * alpha),
        a0: A + 1 + (A - 1) * cos + 2 * sqrtA * alpha,
        a1: -2 * (A - 1 + (A + 1) * cos),
        a2: A + 1 + (A - 1) * cos - 2 * sqrtA * alpha,
      };
    case "highShelf":
      return {
        b0: A * (A + 1 + (A - 1) * cos + 2 * sqrtA * alpha),
        b1: -2 * A * (A - 1 + (A + 1) * cos),
        b2: A * (A + 1 + (A - 1) * cos - 2 * sqrtA * alpha),
        a0: A + 1 - (A - 1) * cos + 2 * sqrtA * alpha,
        a1: 2 * (A - 1 - (A + 1) * cos),
        a2: A + 1 - (A - 1) * cos - 2 * sqrtA * alpha,
      };
    case "lowPass":
      return {
        b0: (1 - cos) / 2,
        b1: 1 - cos,
        b2: (1 - cos) / 2,
        a0: 1 + alpha,
        a1: -2 * cos,
        a2: 1 - alpha,
      };
    case "highPass":
      return {
        b0: (1 + cos) / 2,
        b1: -(1 + cos),
        b2: (1 + cos) / 2,
        a0: 1 + alpha,
        a1: -2 * cos,
        a2: 1 - alpha,
      };
  }
}

/** |H(e^jw)| in dB. The a0 term stays in the denominator, so no normalisation. */
function magnitudeDb(c: Biquad, hz: number, fs: number): number {
  const w = (2 * Math.PI * Math.min(hz, fs / 2)) / fs;
  const cos1 = Math.cos(w);
  const sin1 = Math.sin(w);
  const cos2 = Math.cos(2 * w);
  const sin2 = Math.sin(2 * w);

  const numRe = c.b0 + c.b1 * cos1 + c.b2 * cos2;
  const numIm = -(c.b1 * sin1 + c.b2 * sin2);
  const denRe = c.a0 + c.a1 * cos1 + c.a2 * cos2;
  const denIm = -(c.a1 * sin1 + c.a2 * sin2);

  const num = Math.hypot(numRe, numIm);
  const den = Math.hypot(denRe, denIm);
  if (den === 0) return 0;
  // Floored rather than allowed to reach -Infinity, which no plot can draw.
  return Math.max(-90, 20 * Math.log10(Math.max(num / den, 1e-6)));
}

/** One band's contribution, for drawing it on its own when selected. */
export function bandGainAt(band: EqBand, hz: number, fs = DEFAULT_SAMPLE_RATE): number {
  return magnitudeDb(coefficients(band, fs), hz, fs);
}

/**
 * The combined response. Built once per edit and then queried per band per
 * frame, which is why the coefficients are computed up front.
 */
export class EqCurve {
  private sections: Biquad[];

  constructor(
    bands: EqBand[],
    private fs: number = DEFAULT_SAMPLE_RATE,
  ) {
    this.sections = bands.map((b) => coefficients(b, fs));
  }

  get isFlat(): boolean {
    return this.sections.length === 0;
  }

  /** Total gain in dB at a frequency — cascaded filters add in dB. */
  gainAt(hz: number): number {
    let total = 0;
    for (const section of this.sections) total += magnitudeDb(section, hz, this.fs);
    return total;
  }
}

File: engine/src/dsp/eq.rs
//! Parametric EQ over band magnitudes.
//!
//! No audio is filtered here. The analyser has already reduced the signal to a
//! magnitude per band, so the EQ is applied as a gain curve sampled at each
//! band's centre frequency — which is what a filter's magnitude response *is*.
//!
//! The shape is therefore the entire specification, and it is the RBJ cookbook
//! biquad rather than an invented bell, so the curve the editor draws and the
//! gain applied here come from the same closed form. `ui/src/config/eq.ts` is
//! the other half of that pair; the two are asserted against the same
//! properties on both sides.
//!
//! # Where this sits in the chain
//!
//! After AGC, before the range map. Both ends matter:
//!
//! - **After AGC**, because the AGC pushes every band toward a reference level.
//!   An EQ applied before it would be treated as drift and quietly undone over
//!   the AGC time constant.
//! - **Before the range map**, because that is what keeps a boost from lifting
//!   silence. A silent band sits far below the dB floor; adding 12 dB leaves it
//!   far below the dB floor. Applied to the normalised level instead, where
//!   zero *is* the floor, the same boost would make an idle strip glow.

use serde::{Deserialize, Serialize};

pub const Q_MIN: f32 = 0.1;
pub const Q_MAX: f32 = 12.0;

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum EqType {
    /// Peaking bell.
    #[default]
    Peak,
    LowShelf,
    HighShelf,
    LowPass,
    HighPass,
}

impl EqType {
    /// The pass filters cut rather than gain, so their `gain` field is ignored.
    pub fn has_gain(self) -> bool {
        matches!(self, EqType::Peak | EqType::LowShelf | EqType::HighShelf)
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", default)]
pub struct EqBand {
    #[serde(rename = "type")]
    pub kind: EqType,
    pub hz: f32,
    /// dB. Ignored when [`EqType::has_gain`] is false.
    pub gain: f32,
    pub q: f32,
}

impl Default for EqBand {
    fn default() -> Self {
        Self { kind: EqType::Peak, hz: 1000.0, gain: 0.0, q: 1.0 }
    }
}

/// Unnormalised biquad coefficients. `a0` is kept rather than divided out
/// because the magnitude is a ratio of two polynomials and needs it.
#[derive(Clone, Copy, Debug)]
struct Coefficients {
    b0: f32,
    b1: f32,
    b2: f32,
    a0: f32,
    a1: f32,
    a2: f32,
}

/// RBJ Audio EQ Cookbook.
fn coefficients(band: &EqBand, sample_rate: f32) -> Coefficients {
    let nyquist = sample_rate / 2.0;
    let f0 = band.hz.clamp(1.0, nyquist * 0.999);
    let q = band.q.clamp(Q_MIN, Q_MAX);
    let gain = if band.kind.has_gain() { band.gain } else { 0.0 };

    let w0 = std::f32::consts::TAU * f0 / sample_rate;
    let (sin, cos) = w0.sin_cos();
    let alpha = sin / (2.0 * q);
    let a = 10f32.powf(gain / 40.0);
    let sqrt_a = a.sqrt();

    match band.kind {
        EqType::Peak => Coefficients {
            b0: 1.0 + alpha * a,
            b1: -2.0 * cos,
            b2: 1.0 - alpha * a,
            a0: 1.0 + alpha / a,
            a1: -2.0 * cos,
            a2: 1.0 - alpha / a,
        },
        EqType::LowShelf => Coefficients {
            b0: a * (a + 1.0 - (a - 1.0) * cos + 2.0 * sqrt_a * alpha),
            b1: 2.0 * a * (a - 1.0 - (a + 1.0) * cos),
            b2: a * (a + 1.0 - (a - 1.0) * cos - 2.0 * sqrt_a * alpha),
            a0: a + 1.0 + (a - 1.0) * cos + 2.0 * sqrt_a * alpha,
            a1: -2.0 * (a - 1.0 + (a + 1.0) * cos),
            a2: a + 1.0 + (a - 1.0) * cos - 2.0 * sqrt_a * alpha,
        },
        EqType::HighShelf => Coefficients {
            b0: a * (a + 1.0 + (a - 1.0) * cos + 2.0 * sqrt_a * alpha),
            b1: -2.0 * a * (a - 1.0 + (a + 1.0) * cos),
            b2: a * (a + 1.0 + (a - 1.0) * cos - 2.0 * sqrt_a * alpha),
            a0: a + 1.0 - (a - 1.0) * cos + 2.0 * sqrt_a * alpha,
            a1: 2.0 * (a - 1.0 - (a + 1.0) * cos),
            a2: a + 1.0 - (a - 1.0) * cos - 2.0 * sqrt_a * alpha,
        },
        EqType::LowPass => Coefficients {
            b0: (1.0 - cos) / 2.0,
            b1: 1.0 - cos,
            b2: (1.0 - cos) / 2.0,
            a0: 1.0 + alpha,
            a1: -2.0 * cos,
            a2: 1.0 - alpha,
        },
        EqType::HighPass => Coefficients {
            b0: (1.0 + cos) / 2.0,
            b1: -(1.0 + cos),
            b2: (1.0 + cos) / 2.0,
            a0: 1.0 + alpha,
            a1: -2.0 * cos,
            a2: 1.0 - alpha,
        },
    }
}

/// `20·log10 |H(e^jw)|`, floored rather than allowed to reach negative infinity.
fn magnitude_db(c: &Coefficients, hz: f32, sample_rate: f32) -> f32 {
    let w = std::f32::consts::TAU * hz.min(sample_rate / 2.0) / sample_rate;
    let (sin1, cos1) = w.sin_cos();
    let (sin2, cos2) = (2.0 * w).sin_cos();

    let num = (c.b0 + c.b1 * cos1 + c.b2 * cos2).hypot(-(c.b1 * sin1 + c.b2 * sin2));
    let den = (c.a0 + c.a1 * cos1 + c.a2 * cos2).hypot(-(c.a1 * sin1 + c.a2 * sin2));
    if den == 0.0 {
        return 0.0;
    }
    (20.0 * (num / den).max(1e-6).log10()).max(-90.0)
}

/// One band's contribution on its own, for tests and for the editor's solo view.
pub fn band_gain_db(band: &EqBand, hz: f32, sample_rate: f32) -> f32 {
    magnitude_db(&coefficients(band, sample_rate), hz, sample_rate)
}

/// The combined response of a cascade.
#[derive(Clone, Debug, Default)]
pub struct EqCurve {
    sections: Vec<Coefficients>,
    sample_rate: f32,
}

impl EqCurve {
    pub fn new(bands: &[EqBand], sample_rate: f32) -> Self {
        Self {
            sections: bands.iter().map(|b| coefficients(b, sample_rate)).collect(),
            sample_rate,
        }
    }

    pub fn is_flat(&self) -> bool {
        self.sections.is_empty()
    }

    /// Total gain in dB. Cascaded filters multiply in magnitude, so they add here.
    pub fn gain_db(&self, hz: f32) -> f32 {
        self.sections
            .iter()
            .map(|c| magnitude_db(c, hz, self.sample_rate))
            .sum()
    }

    /// Pre-sampled at each band centre, so the per-frame cost is a slice read.
    pub fn table(&self, centers: &[f32]) -> Vec<f32> {
        if self.is_flat() {
            return vec![0.0; centers.len()];
        }
        centers.iter().map(|&f| self.gain_db(f)).collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const SR: f32 = 48_000.0;

    fn band(kind: EqType, hz: f32, gain: f32, q: f32) -> EqBand {
        EqBand { kind, hz, gain, q }
    }

    /// Closed-form properties, not values captured from a previous run: a
    /// transcription slip in the coefficients fails these rather than being
    /// baked in as the new expectation.
    #[test]
    fn bell_reaches_its_full_gain_at_centre() {
        // |H(w0)| = A^2 for the peaking filter, i.e. exactly the requested dB.
        for gain in [-18.0, -6.0, 3.0, 12.0] {
            let g = band_gain_db(&band(EqType::Peak, 1000.0, gain, 1.4), 1000.0, SR);
            assert!((g - gain).abs() < 1e-3, "asked {gain} dB, got {g}");
        }
    }

    #[test]
    fn bell_leaves_distant_bands_alone() {
        let b = band(EqType::Peak, 1000.0, 12.0, 2.0);
        assert!(band_gain_db(&b, 40.0, SR).abs() < 0.05);
        assert!(band_gain_db(&b, 18_000.0, SR).abs() < 0.2);
    }

    #[test]
    fn lower_q_widens_the_bell() {
        let at = |q| band_gain_db(&band(EqType::Peak, 1000.0, 12.0, q), 500.0, SR);
        assert!(at(0.3) > at(1.0));
        assert!(at(1.0) > at(6.0));
    }

    #[test]
    fn shelves_lift_one_side_only() {
        let low = band(EqType::LowShelf, 500.0, 9.0, 0.707);
        assert!((band_gain_db(&low, 25.0, SR) - 9.0).abs() < 0.1);
        assert!(band_gain_db(&low, 15_000.0, SR).abs() < 0.1);

        let high = band(EqType::HighShelf, 2000.0, -9.0, 0.707);
        assert!((band_gain_db(&high, 15_000.0, SR) + 9.0).abs() < 0.1);
        assert!(band_gain_db(&high, 25.0, SR).abs() < 0.1);
    }

    #[test]
    fn pass_filters_are_three_db_down_at_cutoff() {
        // |H(w0)| = Q for both pass types, so Q = 1/sqrt(2) is the -3.01 dB point.
        let q = std::f32::consts::FRAC_1_SQRT_2;
        for kind in [EqType::LowPass, EqType::HighPass] {
            let g = band_gain_db(&band(kind, 1000.0, 0.0, q), 1000.0, SR);
            assert!((g + 3.0103).abs() < 1e-2, "{kind:?} was {g} dB at cutoff");
        }
    }

    #[test]
    fn pass_filters_cut_the_far_side() {
        let q = std::f32::consts::FRAC_1_SQRT_2;
        let low = band(EqType::LowPass, 1000.0, 0.0, q);
        assert!(band_gain_db(&low, 100.0, SR).abs() < 0.05);
        assert!(band_gain_db(&low, 8000.0, SR) < -30.0);

        let high = band(EqType::HighPass, 1000.0, 0.0, q);
        assert!(band_gain_db(&high, 8000.0, SR).abs() < 0.05);
        assert!(band_gain_db(&high, 100.0, SR) < -30.0);
    }

    #[test]
    fn cascaded_bands_add_in_db() {
        let a = band(EqType::Peak, 200.0, 6.0, 1.0);
        let b = band(EqType::Peak, 5000.0, -4.0, 1.0);
        let curve = EqCurve::new(&[a.clone(), b.clone()], SR);
        for hz in [50.0, 200.0, 1000.0, 5000.0, 16_000.0] {
            let expected = band_gain_db(&a, hz, SR) + band_gain_db(&b, hz, SR);
            assert!((curve.gain_db(hz) - expected).abs() < 1e-4);
        }
    }

    #[test]
    fn empty_cascade_is_flat() {
        let curve = EqCurve::new(&[], SR);
        assert!(curve.is_flat());
        assert_eq!(curve.gain_db(1000.0), 0.0);
        assert_eq!(curve.table(&[100.0, 1000.0]), vec![0.0, 0.0]);
    }

    /// The editor's axis reaches 20 kHz, which is above Nyquist at 32 kHz. A
    /// band parked up there must not produce NaN and poison every level.
    #[test]
    fn survives_a_band_above_nyquist() {
        for kind in [EqType::Peak, EqType::LowPass, EqType::HighShelf] {
            let g = band_gain_db(&band(kind, 20_000.0, 12.0, 4.0), 20_000.0, 32_000.0);
            assert!(g.is_finite(), "{kind:?} produced {g}");
        }
    }

    #[test]
    fn q_outside_the_usable_range_is_clamped_not_exploded() {
        for q in [0.0, -1.0, 1e9] {
            let g = band_gain_db(&band(EqType::Peak, 1000.0, 6.0, q), 1000.0, SR);
            assert!(g.is_finite(), "Q {q} produced {g}");
        }
    }

    #[test]
    fn parses_the_editor_wire_format() {
        let json = r#"{"id":"eq-1","type":"highShelf","hz":2000,"gain":-6,"q":0.7}"#;
        let band: EqBand = serde_json::from_str(json).unwrap();
        assert_eq!(band.kind, EqType::HighShelf);
        assert_eq!(band.hz, 2000.0);
        assert_eq!(band.gain, -6.0);
    }
}
